contrib/masterfiles-stage/common.sh

```diff
@@ -12,146 +12,244 @@ if [ ! "$debug_mode" == "true" ]; then
+    git --git-dir="${local_mirrored_repo}" fetch && return 0
+
+    # If execution arrives here, the local_mirrored_repo exists but is messed up somehow
+    # (or there is network trouble).  Easiest is to wipe it and start fresh.
+    rm -rf "${local_mirrored_repo}"
   fi
+  git clone --mirror "${GIT_URL}" "${local_mirrored_repo}" ||
+    error_exit "Failed: git clone --mirror '${GIT_URL}' '${local_mirrored_repo}'"
+}
+
+git_stage_refspec() {
+  # This function depends on git_setup_local_mirrored_repo
+  # having been run, such that the variable local_mirrored_repo
+  # contains the (local) path to a bare git repository.
+  #
+  # (A mirror repository is a special case of a bare repository;
+  # either will work for this function but a bare non-mirror
+  # repository will have edge cases that are mishandled.)
+  #
+  # This function accepts a single argument: refspec,
+  # which should be a git tagname, branch, or commit hash.
+  #
+  # This function stages the refspec to the STAGING_DIR
+  # from local_mirrored_repo
+
+  mkdir -p "${STAGING_DIR}" || error_exit "Failed: mkdir -p '$STAGING_DIR'"
+  git --git-dir="${local_mirrored_repo}" --work-tree="${STAGING_DIR}" checkout -q -f "$1" ||
+    error_exit "Failed to checkout '$2' from '${local_mirrored_repo}'"
+  git --git-dir="${local_mirrored_repo}" --work-tree="${STAGING_DIR}" clean -q -dffx ||
+    error_exit "Failed: git --git-dir='${local_mirrored_repo}' --work-tree='${STAGING_DIR}' clean -q -dffx"
+}
+
+validate_staged_policy() {
+  # If you use this function, ensure you have set STAGING_DIR.
+  # Also see function "avoid_triggering_unneeded_policy_updates"
+  /var/cfengine/bin/cf-promises -T "${STAGING_DIR}" &&
+  /var/cfengine/bin/cf-promises -cf "${STAGING_DIR}/update.cf" ||
+  error_exit "Update policy staged in ${STAGING_DIR} could not be validated, aborting."
 }
 
-git_tag_or_commit_masterstage() {
-  # This function is designed to stage masterfiles from a git TAG or COMMIT
-  #   - Ensure git checkout exists, if it does not remove
-  #     obstructions and create new clone
-  #   - Ensure the origin is set properly
-  #   - Fetch updates
-  #   - Stash any local changes
-  #   - Checkout the proper branch
-  #   - Hard reset to remove any changes
-  #   - Validate the staged policy
-  #   - If Staged policy has validated and the release IDs are different
-  #     synchronize to final location (should we sync anyway to make sure the
-  #     distribution point is good?)
-
-  if ! type "git" > /dev/null; then
-    error_exit "git not found on path: ${PATH}"
+avoid_triggering_unneeded_policy_updates() {
+  # cf_promises_validated gets updated by any run of cf-promises,
+  # but hosts use cf_promises_validated as the flag file to see
+  # if they need to update everything else (the full policy set.)
+  #
+  # cf_promises_release_id is the same for a given policy set
+  # unless changes have actually been made to the policy, so it
+  # can be used to check if we want to trigger an update.
```

Comment:
@nickanderson, is there a reason why [the update policy](https://github.com/cfengine/masterfiles/blob/master/cfe_internal/update/update_policy.cf#L107-L114) uses `cf_promises_validated` rather than `cf_promises_release_id` as the flag file to trigger a full policy set download?